spring-integration-ip/src/main/java/org/springframework/integration/ip/udp/DatagramPacketMessageMapper.java

```diff
@@ -142,7 +145,23 @@ public DatagramPacket fromMessage(Message<?> message) throws Exception {
 			buffer.put(bytes);
 			bytes = buffer.array();
 		}
-		return new DatagramPacket(bytes, bytes.length);
+		return getDatagramFromHeader(message, bytes, bytes.length);
+	}
+
+	private DatagramPacket getDatagramFromHeader(Message<?> message, byte[] bytes,
+												 int length) throws UnknownHostException {
+		InetAddress inetAddress = null;
+		if (message.getHeaders().get(IpHeaders.IP_ADDRESS, String.class) != null) {
+			inetAddress = InetAddress.getByName(message.getHeaders()
+					.get(IpHeaders.IP_ADDRESS, String.class));
```

Comment:
@artembilan I don't quite feel why we should add `enableHeaderRouting`-like option? Using address and port from header is necessary when using `socket-expression` so when user would want to use external socket he also had to use `enableHeaderRouting` set to `true` mandatory. Otherwise sending adapter will not know where to send messages because when using `socket-expression` it is forbidden to use `host` and `port` options (`host` and `port` options in outbound adapter indicate where to send datagrams).
Or maybe I'm missing something here?
@garyrussell Your thoughts about this one?